staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go

```diff
@@ -255,42 +253,12 @@ func TestTimeoutHeaders(t *testing.T) {
 	res.Body.Close()
 }
 
-func captureStdErr() (func() string, func(), error) {
-	var buf bytes.Buffer
-	reader, writer, err := os.Pipe()
-	if err != nil {
-		return nil, nil, err
-	}
-	stderr := os.Stderr
-	readerClosedCh := make(chan struct{})
-	stopReadingStdErr := func() string {
-		writer.Close()
-		<-readerClosedCh
-		return buf.String()
-	}
-	klog.LogToStderr(true)
-	cleanUp := func() {
-		os.Stderr = stderr
-		klog.LogToStderr(false)
-		stopReadingStdErr()
-	}
-	os.Stderr = writer
-	go func() {
-		io.Copy(&buf, reader)
-		readerClosedCh <- struct{}{}
-		close(readerClosedCh)
-	}()
-	klog.LogToStderr(true)
-
-	return stopReadingStdErr, cleanUp, nil
-}
-
 func TestErrConnKilled(t *testing.T) {
-	readStdErr, cleanUp, err := captureStdErr()
-	if err != nil {
-		t.Fatalf("unable to setup the test, err %v", err)
-	}
-	defer cleanUp()
+	buf := new(buffer)
+	klog.SetOutput(buf)
+	klog.LogToStderr(false)
+	defer klog.LogToStderr(true)
```

Comment:
This looks much better than redirecting os.Stderr.

The only problem is that klog has no API to undo `klog.SetOutput(buf)` but as it isn't used when output gets sent again to os.Stderr exclusively, that shouldn't be a problem for other tests.

Reply:
I was confused as this data race was caused by tripping up with two test cases, `TestTimeoutHeaders` and `TestErrConnKilled`. The trace log can confirm this fact: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/108541/pull-kubernetes-unit/1500141824923144192
AFAIK, go runs unit tests in serial, why should this happen? Do you have any ideas?

Reply:
Does the test clean up its own goroutines properly?

In that data race, goroutine 588 calls klog.Error:
```
Previous read at 0x00000388c800 by goroutine 588:
  k8s.io/kubernetes/vendor/k8s.io/klog/v2.(*loggingT).output()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:1036 +0x73e
  k8s.io/kubernetes/vendor/k8s.io/klog/v2.(*loggingT).printDepth()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:735 +0x2d4
  k8s.io/kubernetes/vendor/k8s.io/klog/v2.ErrorDepth()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:1584 +0x104
  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.logError()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:115 +0x1d
  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleError()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:109 +0xa7
  k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP.func2.1()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:137 +0x1d9
```

It was created here:
```
Goroutine 588 (running) created at:
  k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP.func2()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:124 +0xdc
  k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:144 +0x452
  k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.TestTimeoutHeaders.func2.1()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout_test.go:218 +0x371
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2047 +0x4d
  net/http.serverHandler.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2879 +0x89a
  net/http.(*conn).serve()
      /usr/local/go/src/net/http/server.go:1930 +0x12e4
  net/http.(*Server).Serve·dwrap·87()
      /usr/local/go/src/net/http/server.go:3034 +0x58
```

That looks like there is still some background activity going on, either from the same test or some other test, while this code gets executed and changes os.Stderr:
```
	cleanUp := func() {
		os.Stderr = stderr
		klog.LogToStderr(false)
		stopReadingStdErr()
	}
```

Reply:
This maybe the background goroutine https://github.com/kubernetes/apiserver/blob/4c7efc059ce67e25680c1edc42c5f438367fd60e/pkg/server/filters/timeout.go#L120-L139 which handles the incoming requests.
In this situation, we known `klog.output` has mutex protected, but I think it's useless. Klog still works well but the result was hijacked by another thread. In this test, we redirected the output without using klog, so we needed to lock the io by ourself. what do you think?

Reply:
If klog keeps writing into the buffer, then you would need to protect read access.

Another, simpler solution is to replace the buffer one more time before reading it.

Reply:
The buffer checking also must be flexible enough to work when some runaway goroutine keeps writing to the buffer. It would be better to stop that background goroutine.

Reply:
> Another, simpler solution is to replace the buffer one more time before reading it.

I want to hijack the result the klog outputs, I think I cant't simply do something like `klog.SetBuffer`, if so, I will miss some message. If I understood right?

Reply:
Does it busy poll?

Reply:
> You do that once you are sure that the expected output must have been emitted. If you cannot be sure of that, how can the test verify the log output?

Make sense then. Updated.